osf/utils/machines.py

```diff
@@ -344,118 +346,127 @@ def get_context(self):
-        pass
+    @property
+    def target_registration(self):
+        raise NotImplementedError(
+            'SanctionStateMachine subclasses must define a target_registration property'
+        )
 
-    def notify_withdraw_request_submitted(self, ev):
-        pass
+    @property
+    def approval_stage(self):
+        raise NotImplementedError(
+            'SanctionStateMachine subclasses must define an approval_stage property with a setter.'
+        )
 
-    def notify_withdraw_request_denied(self, ev):
-        pass
+    def _process(self, *args, **kwargs):
+        '''Wrap superclass _process to handle expected MachineErrors.'''
+        try:
+            super()._process(*args, **kwargs)
+        except MachineError as e:
+            short_message = 'Operation not allowed at this time'
+            if self.approval_stage in [SanctionStates.ADMIN_REJECTED, SanctionStates.MODERATOR_REJECTED]:
+                long_message = (
+                    'This {sanction} has already been rejected and cannot be approved'.format(
+                        sanction=self.DISPLAY_NAME))
+            elif self.approval_stage in [SanctionStates.ACCEPTED, SanctionStates.COMPLETE]:
+                long_message = (
+                    'This {sanction} has all required approvals and cannot be rejected'.format(
+                        sanction=self.DISPLAY_NAME))
+            else:
+                raise e
+
+            raise HTTPError(http_status.HTTP_400_BAD_REQUEST, data={
+                'message_short': short_message, 'message_long': long_message})
+
+    def _initialize_transition(self, event_data):
+        self.from_state = self.target_registration.moderation_state
+
+    def _infer_last_trigger(self):
+        '''Determine the most recent 'trigger' for the sanction from a moderation perspective.'''
+        if self.SANCTION_TYPE is SanctionTypes.RETRACTION:
+            if self.approval_stage is SanctionStates.MODERATOR_REJECTED:
+                return RegistrationModerationTriggers.REJECT_WITHDRAWAL
+            elif self.approval_stage is SanctionStates.ACCEPTED:
+                return RegistrationModerationTriggers.ACCEPT_WITHDRAWAL
+            else:  # PENDING_ADMIN_APPROVAL or PENDING_MODERATOR_APPROVAL
+                return RegistrationModerationTriggers.REQUEST_WITHDRAWAL
+
+        if self.SANCTION_TYPE in [SanctionTypes.EMBARGO, SanctionTypes.REGISTRATION_APPROVAL]:
+            if self.approval_stage is SanctionStates.MODERATOR_REJECTED:
+                return RegistrationModerationTriggers.REJECT_SUBMISSION
+            elif self.approval_stage is SanctionStates.ACCEPTED:
+                return RegistrationModerationTriggers.ACCEPT_SUBMISSION
+            else:  # PENDING_ADMIN_APPROVAL or PENDING_MODERATOR_APPROVAL
+                return RegistrationModerationTriggers.SUBMIT
+
+        return None  # No moderation triggers for other sanction types
+
+    def _log_moderated_action(self, event_data):
```

Comment:
I don't love the SanctionStateMachine knowing about RegistrationActions. I'm thinking this function may belong on the Registration. Could possibly even happen as part of update_moderation_state?